src/libraries/System.Text.Json/src/System/Text/Json/Serialization/Converters/Object/ObjectConverterFactory.cs

```diff
@@ -21,16 +22,119 @@ public override bool CanConvert(Type typeToConvert)
+                int parameterCount = parameters.Length;
+
+                if (parameterCount <= JsonConstants.UnboxedParameterCountThreshold)
+                {
+                    Type placeHolderType = typeof(object);
+                    Type[] typeArguments = new Type[JsonConstants.UnboxedParameterCountThreshold + 1];
+
+                    typeArguments[0] = typeToConvert;
+                    for (int i = 0; i < JsonConstants.UnboxedParameterCountThreshold; i++)
+                    {
+                        if (i < parameterCount)
+                        {
+                            typeArguments[i + 1] = parameters[i].ParameterType;
+                        }
+                        else
+                        {
+                            // Use placeholder arguments if there are less args than the threshold.
+                            typeArguments[i + 1] = placeHolderType;
+                        }
+                    }
+
+                    converterType = typeof(SmallObjectWithParameterizedConstructorConverter<,,,,>).MakeGenericType(typeArguments);
+                }
+                else
+                {
+                    converterType = typeof(LargeObjectWithParameterizedConstructorConverter<>).MakeGenericType(typeToConvert);
+                }
+            }
+
+            converter = (JsonConverter)Activator.CreateInstance(
+                    converterType,
+                    BindingFlags.Instance | BindingFlags.Public,
+                    binder: null,
+                    args: null,
+                    culture: null)!;
+
+            converter.ConstructorInfo = constructor!;
             return converter;
         }
+
+        private ConstructorInfo? GetDeserializationConstructor(Type type)
+        {
+            ConstructorInfo? ctorWithAttribute = null;
+            ConstructorInfo? publicParameterlessCtor = null;
+            ConstructorInfo? lonePublicCtor = null;
+
+            // Check public constructors.
+            ConstructorInfo[] constructors = type.GetConstructors(BindingFlags.Public | BindingFlags.Instance);
+
+            if (constructors.Length == 1)
+            {
+                lonePublicCtor = constructors[0];
+            }
+
+            foreach (ConstructorInfo constructor in constructors)
+            {
+                if (constructor.GetCustomAttribute<JsonConstructorAttribute>() != null)
+                {
+                    if (ctorWithAttribute != null)
+                    {
+                        ThrowHelper.ThrowInvalidOperationException_SerializationDuplicateTypeAttribute<JsonConstructorAttribute>(type);
+                    }
+
+                    ctorWithAttribute = constructor;
+                }
+                else if (publicParameterlessCtor == null && constructor.GetParameters().Length == 0)
+                {
+
+                    publicParameterlessCtor = constructor;
+                }
+            }
+
+            // Now, check non-public ctors.
```

Comment:
Yes, I left it in for some time in case there was any pushback. I've removed the support now.

I left  in logic to throw if there are multiple ctors with the `JsonConstructor` attribute (even if one or more are non-public), to avoid first-ctor-with-attribute-wins semantics if we add non-public support in the future (that would be to avoid breaking behavior where we go from not throwing to throwing.)